ion/binarywriter.go

```diff
@@ -251,22 +249,51 @@ func (w *binaryWriter) WriteTimestamp(val Timestamp) error {
 	return w.writeValue("Writer.WriteTimestamp", buf)
 }
 
-// WriteSymbol writes a symbol value.
-func (w *binaryWriter) WriteSymbol(val string) error {
-	id, err := w.resolve("Writer.WriteSymbol", val)
-	if err != nil {
-		w.err = err
-		return err
+// WriteSymbol writes a symbol value given a SymbolToken.
+func (w *binaryWriter) WriteSymbol(val SymbolToken) error {
+	var id uint64
+	if val.LocalSID != SymbolIDUnknown {
+		id = uint64(val.LocalSID)
+	} else if val.Text != nil {
+		text := *val.Text
+		if _, ok := symbolIdentifier(text); ok {
+			// Wrap text value in single quotes if the symbol's text is a symbol identifier
+			// (ie. of form $n for some integer n)
+			// This is done to distinguish from actual symbol table mappings.
+			text = fmt.Sprintf("'%v'", text)
+		}
+
+		id, w.err = w.resolveFromSymbolTable("Writer.WriteSymbol", text)
+		if w.err != nil {
+			return w.err
+		}
+	} else {
+		return &UsageError{"Writer.WriteSymbol", "invalid symbol token"}
+	}
+
+	return w.writeSymbolFromID("Writer.WriteSymbol", id)
+}
+
+// WriteSymbolFromString writes a symbol value given a string.
```

Comment:
This comment should be more explicit about how the provided string is going to be used; it expects the string to already be in the symbol table and will return an error if it isn't.

Reply:
Done.